pkg/mutate/squashfs.go

```diff
@@ -104,23 +123,57 @@ func SquashfsLayer(base v1.Layer, dir string, opts ...SquashfsConverterOpt) (v1.
 }
 
 // makeSquashfs returns the path to a squashfs file that contains the contents of the uncompressed
-// TAR stream from r.
-func (c *squashfsConverter) makeSquashfs(r io.Reader) (string, error) {
+// TAR stream from l.Uncompressed().
+func (c *squashfsConverter) makeSquashfs(l v1.Layer) (string, error) {
 	dir, err := os.MkdirTemp(c.dir, "")
 	if err != nil {
 		return "", err
 	}
 
 	path := filepath.Join(dir, "layer.sqfs")
 
+	tarIn, err := l.Uncompressed()
+	if err != nil {
+		return "", err
+	}
+	defer tarIn.Close()
+
+	var opaquePaths map[string]bool
+	filterErr := make(chan error, 1)
+
+	if c.convertWhiteout {
+		opqrc, err := l.Uncompressed()
+		if err != nil {
+			return "", err
+		}
+		defer opqrc.Close()
+		opaquePaths, err = scanAUFSOpaque(opqrc)
+		if err != nil {
+			return "", err
+		}
+
+		filterIn := tarIn
+		pr, pw := io.Pipe()
+		tarIn = pr
+		go func() {
+			filterErr <- whiteoutFilter(filterIn, pw, opaquePaths)
+		}()
```

Comment:
If `len(opaquePaths) == 0`, could we could skip this step as an optimization?

Reply:
Not as written, given opaquePaths is only for the directory `.wh..wh..opq` markers. We might still have `.wh.<file>` entries to process in the filter.

We could make the scan look for `.wh.<file>` markers as well and then skip the filter if there are none of those, and there are also no opaque directory markers.